subxt/src/error/mod.rs

```diff
@@ -162,3 +162,10 @@ pub enum StorageAddressError {
         fields: usize,
     },
 }
+
+/// This trait is implemented on the statically generated root ModuleError type
+#[doc(hidden)]
+pub trait RootError: Sized {
+    /// Given details of the pallet error we want to decode
+    fn root_error(pallet_index: &u8, error: &[u8; 4]) -> Result<Self, Error>;
+}
```

Comment:
I'd more closely mirror the events version which looks like:

```rust
pub trait RootEvent: Sized {
    fn root_event(
        pallet_bytes: &[u8],
        pallet_name: &str,
        pallet_event_ty: u32,
        metadata: &Metadata,
    ) -> Result<Self, Error>;
}
```

So something more like this maybe:

```rust
pub trait RootError: Sized {
    fn root_error(
        pallet_bytes: [u8; 4],
        pallet_name: &str,
        pallet_event_ty: u32,
        metadata: &Metadata,
    ) -> Result<Self, Error>;
}
```

Reasons:
1. We should avoid relying on `Decode::decode` if we can and instead use `#mod_name::Error::decode_with_metadata` to decode it. Why? because `Decode::decode` will just decode the bytes based on the static codegenned type, whereas `decode_with_metadata` will decode into the error enum based on matching up the encoded variant names and fields with the error struct we've codegenned, making it more robust to changes in eg variant order or field order or whatever.
2. Using a string instead of index for pallet; same as above, just allows more robust matching across nodes where pallets are likely in a different order.
3. Just generally to be consistent with RootError :)